quarkchain/evm/messages.py

```diff
@@ -350,6 +363,28 @@ def apply_transaction(state, tx: transactions.Transaction, tx_wrapper_hash):
         state.get_balance(tx.sender, token_id=tx.gas_token_id)
         >= tx.startgas * tx.gasprice
     )
+    gasprice, refund_rate = tx.gasprice, 100
+    # convert gas if using non-genesis native token
+    if tx.gas_token_id != state.genesis_token:
+        refund_rate, converted_genesis_token_gas_price = pay_native_token_as_gas(
+            state, tx.gas_token_id, tx.startgas, tx.gasprice
+        )
+        assert converted_genesis_token_gas_price > 0
```

Comment:
if contract call to `payAsGas` failed, should we mark this transaction failed? @qizhou